File: Chapter.14/StockTicker.Server.CSharp/Core/TradingCoordinator.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.AspNetCore.SignalR;
using ReactiveAgent.CSharp;
using StockTicker.Core;
using TwitterEmotionAnalysis.CSharp.RxPublisherSubscriber;

namespace StockTicker.Server.CSharp.Core
{
    using static TradingSupervisorAgent;

    public class TradingCoordinator : IDisposable
    {
        private static readonly Lazy<TradingCoordinator> tradingCoordinator =
            new Lazy<TradingCoordinator>(() => new TradingCoordinator());

        private readonly IAgent<CoordinatorMessage> coordinatorAgent;

        private readonly RxPubSub<Models.Trading> subject;

        private TradingCoordinator()
        {
            subject = new RxPubSub<Models.Trading>();
            coordinatorAgent =
                Agent.Start<Dictionary<string, (IObserver<Models.Trading>, IDisposable)>,
                    CoordinatorMessage>(
                    new Dictionary<string, (IObserver<Models.Trading>, IDisposable)>(),
                    async (agents, message) =>
                    {
                        switch (message)
                        {
                            case CoordinatorMessage.Subscribe msgSub:
                                var observer = new TradingAgent(msgSub.connId, msgSub.userName, msgSub.initialAmount,
                                    msgSub.caller);
                                var dispObsrever = subject.Subscribe(observer);
                                await msgSub.caller.SendAsync("setInitialAsset", msgSub.initialAmount);
                                agents.Add(msgSub.connId, (observer, dispObsrever));
                                break;
                            case CoordinatorMessage.Unsubscribe msgUnsub:
                                if (agents.TryGetValue(msgUnsub.id, out var value))
                                {
                                    value.Item2.Dispose();
                                    agents.Remove(msgUnsub.id);
                                }

                                break;
                            case CoordinatorMessage.PublishCommand msgCmd:
                                var id = msgCmd.connId;
                                if (agents.TryGetValue(id, out var agent))
                                    switch (msgCmd.Item2.Command)
                                    {
                                        case Models.TradingCommand.BuyStockCommand buy:
                                            agent.Item1.OnNext(Models.Trading.NewBuy(buy.tradingRecord.Symbol,
                                                buy.tradingRecord));

                                            break;
                                        case Models.TradingCommand.SellStockCommand sell:
                                            agent.Item1.OnNext(Models.Trading.NewSell(sell.tradingRecord.Symbol,
                                                sell.tradingRecord));
                                            break;
                                    }

                                break;
                        }

                        return agents;
                    });
        }

        public void Dispose()
        {
            subject.Dispose();
        }

        public Task Subscribe(string connId, string userName, decimal initialAmount, IClientProxy caller)
        {
            return coordinatorAgent.Send(CoordinatorMessage.NewSubscribe(connId, userName, initialAmount, caller));
        }

        public Task Unsubscribe(string connId)
        {
            return coordinatorAgent.Send(CoordinatorMessage.NewUnsubscribe(connId));
        }

        public Task PublishCommand(CoordinatorMessage command)
        {
            return coordinatorAgent.Send(command);
        }

        public IDisposable AddPublisher(IObservable<Models.Trading> observable)
        {
            return subject.AddPublisher(observable);
        }

        public static TradingCoordinator Instance()
        {
            return tradingCoordinator.Value;
        }
    }
}

File: Chapter.14/StockTicker.Server.FSharp/Core/TradingCoordinator.fs
namespace StockTicker.Server.FSharp

module TradingCoordinator = 

    open System
    open Microsoft.AspNetCore.SignalR
    open StockTicker.Core
    open TradingAgent
    open TwitterEmotionAnalysis.CSharp.RxPublisherSubscriber
    
    // responsible for subscribing and un-subscribing TradingAgent
    // it uses a mix of RX and Agent.Post just for demo purpose
    // (TradingAgent : IOboservable) and (TradingSuperviserAgent : IObservable)
    type TradingCoordinator() =   // #A

        //Listing 6.6 Reactive Publisher Subscriber in C#
        let subject = new RxPubSub<Trading>()    // #B
        static let tradingCoordinator =
            System.Lazy<TradingCoordinator>.Create(fun () -> new TradingCoordinator())  // #C

        let coordinatorAgent =
            Agent<CoordinatorMessage>.Start(fun inbox ->
                let rec loop (agents : Map<string, (IObserver<Trading> * IDisposable)>) =
                    async {
                        let! msg = inbox.Receive()
                        match msg with
                        | Subscribe(connId, username, amount, caller) ->    // #D
                            let observer = TradingAgent(connId, username, amount, caller)  // #E                            
                            let dispObsrever = subject.Subscribe(observer)
                            observer.Agent |> withSupervisor connId supervisor |> startAgent   // #F                            
                            do! caller.SendAsync("setInitialAsset", amount) |> Async.AwaitTask    // #G
                            return! loop (Map.add connId (observer :> IObserver<Trading>, dispObsrever) agents)
                        | Unsubscribe(connId) ->
                            match Map.tryFind connId agents with
                            | Some(_, disposable) ->  // #H
                                disposable.Dispose()
                                return! loop (Map.remove connId agents)
                            | None -> return! loop agents
                        | PublishCommand(connId, command) ->   // #I
                            match Map.tryFind connId agents with
                            | Some(agent, _) ->
                                match command.Command with
                                | TradingCommand.BuyStockCommand(id, trading) ->    
                                    agent.OnNext(Trading.Buy(trading.Symbol, trading))
                                    return! loop agents
                                | TradingCommand.SellStockCommand(connId, trading) ->
                                    agent.OnNext(Trading.Sell(trading.Symbol, trading))
                                    return! loop agents
                            | None -> return! loop agents }
                loop (Map.empty))

        member this.Subscribe(connId : string, username:string, initialAmount : decimal, caller:IClientProxy) =
            coordinatorAgent.Post(Subscribe(connId, username, initialAmount, caller)) // #D

        member this.Unsubscribe(id : string) = coordinatorAgent.Post(Unsubscribe(id))

        member this.PublishCommand(command) = coordinatorAgent.Post(command)  // #L

        member this.AddPublisher(observable : IObservable<Trading>) = subject.AddPublisher(observable)  // #M

        static member Instance() = tradingCoordinator.Value  // #C

        interface IDisposable with   // #O
            member x.Dispose() =  subject.Dispose()